plugins/image-prioritizer/class-image-prioritizer-img-tag-visitor.php

```diff
@@ -40,6 +63,92 @@ public function __invoke( OD_Tag_Visitor_Context $context ): bool {
 
 		$xpath = $processor->get_xpath();
 
+		$parent_tag = $this->get_parent_tag_name( $xpath );
+		if ( 'PICTURE' === $parent_tag ) {
+			return true;
+		}
+
+		$this->pre_img_process( $processor, $context, $xpath );
+
+		$this->add_preload_link_for_img( $processor, $context, $xpath );
+
+		return true;
+	}
+
+	/**
+	 * Process a <picture> element.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param OD_HTML_Tag_Processor  $processor HTML tag processor.
+	 * @param OD_Tag_Visitor_Context $context Tag visitor context.
+	 *
+	 * @return bool Whether the tag should be tracked in URL Metrics.
+	 */
+	private function process_picture( OD_HTML_Tag_Processor $processor, OD_Tag_Visitor_Context $context ): bool {
+		$collected_sources = array();
+		$img_xpath         = null;
+
+		// Loop through child tags until we reach the closing </picture> tag.
+		while ( $processor->next_tag() ) {
+			$tag = $processor->get_tag();
+
+			// If we reached the closing </picture> tag, break.
+			if ( 'PICTURE' === $tag && $processor->is_tag_closer() ) {
+				break;
+			}
+
+			// Collect <source> elements.
+			if ( 'SOURCE' === $tag && ! $processor->is_tag_closer() ) {
+				$media = $processor->get_attribute( 'media' );
+				$type  = $processor->get_attribute( 'type' );
+
+				// Ensure that all <source> elements have a type attribute and no media attribute.
+				if ( null !== $media || null === $type ) {
+					return false;
+				}
+
+				$collected_sources[] = array(
+					'srcset'      => $processor->get_attribute( 'srcset' ),
+					'sizes'       => $processor->get_attribute( 'sizes' ),
+					'type'        => $type,
+					'crossorigin' => $this->get_attribute_value( $processor, 'crossorigin' ),
+				);
+			}
+
+			// Process the <img> element within the <picture>.
+			if ( 'IMG' === $tag && ! $processor->is_tag_closer() ) {
+				// Skip empty src attributes and data: URLs.
+				$src = trim( (string) $processor->get_attribute( 'src' ) );
+				if ( '' === $src || $this->is_data_url( $src ) ) {
+					return false;
+				}
+
+				$img_xpath = $processor->get_xpath();
+
+				$this->pre_img_process( $processor, $context, $img_xpath );
+			}
```

Comment:
Do we need this here? Couldn't we just rather leave the existing `IMG` handling in place when `__invoke()` lands on an `IMG` tag? We would just modify the logic there to only add preload links if the parent is not a `PICTURE`.

```suggestion
```

Reply:
I have removed the `pre_img_process` as this processed directly in `process_img` but the below logic is still there as it is needed for images `xpath` which is required for adding preload links in `process_picture`.
```php
if ( 'IMG' === $tag && ! $processor->is_tag_closer() ) {
    // Skip empty src attributes and data: URLs.
    $src = trim( (string) $processor->get_attribute( 'src' ) );
    if ( '' === $src || $this->is_data_url( $src ) ) {
        return false;
    }
    $img_xpath = $processor->get_xpath();
}
```